### PR TITLE
support ast description

### PR DESCRIPTION
**Pull request by olgakil**

### CODE REVIEW

File: src/test/java/com/checkmarx/flow/cucumber/common/utils/AstUtils.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
+package com.checkmarx.flow.cucumber.common.utils;
+
+import com.cx.restclient.ast.dto.sast.report.Finding;
+import org.junit.Assert;
+
+import java.util.*;
+
+public class AstUtils {
+
+    public static void validateDescriptions(List<Finding> findings) {
```
Comment (alex-ko-dev) [inline src/test/java/com/checkmarx/flow/cucumber/common/utils/AstUtils.java:10]: Do we need a separate class for this method? It's only called from one place.

File: src/test/java/com/checkmarx/flow/cucumber/common/utils/AstUtils.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+        Map<String, Set<String>> mapDescriptions = new HashMap<>();
+
+        findings.forEach(finding -> {
+            Set<String> listDescriptions = mapDescriptions.get(finding.getQueryID());
+            if(listDescriptions == null){
+                listDescriptions = new HashSet<>();
+            }
```
Comment (alex-ko-dev) [inline src/test/java/com/checkmarx/flow/cucumber/common/utils/AstUtils.java:18]: Can be replaced with `Optional.ofNullable( mapDescriptions.get(finding.getQueryID())).orElseGet(HashSet::new)`

File: src/test/java/com/checkmarx/flow/cucumber/integration/ast/scans/AstRemoteRepoScanSteps.java [hunk 3]
```diff
@@ -40,9 +40,14 @@
 @Slf4j
 @RequiredArgsConstructor
 public class AstRemoteRepoScanSteps {
+
     final String EMPTY_STRING_INDICATOR = "<empty>";
     private static final String PUBLIC_PROJECT_NAME = "Public-Test-Test-Repo";
 
+    private static final String BRANCH = "master";
+
+    private static final String REPO_DESCRIPTION = "https://github.com/cxflowtestuser/public-rest-repo";
```
Comment (alex-ko-dev) [inline src/test/java/com/checkmarx/flow/cucumber/integration/ast/scans/AstRemoteRepoScanSteps.java:49]: This is a URL, not description.

File: src/test/java/com/checkmarx/flow/cucumber/integration/ast/scans/AstRemoteRepoScanSteps.java [hunk 4]
```diff
@@ -173,8 +178,30 @@ public void astScanWithInvalidAstMachine(String exceptionType, String errorSubst
...
 
+    @Given("enabled vulnerability scanners are {string} and branch is {string}")
+    public void setScanInitiatorAndBranch(String initiatorList, String branch) {
+        List<VulnerabilityScanner> scanners = setScanInitiator(initiatorList);
+        startScan(scanners, branch, REPO_DESCRIPTION);
+    }
+    
+    @And("each finding will contain AST populated description field")
+    public void validateAdditionalFields(){
+        assertTrue("AST scan ID is empty", StringUtils.isNotEmpty(scanResults.getAstResults().getResults().getFindings().get(0).getDescription()));
```
Comment (alex-ko-dev) [inline src/test/java/com/checkmarx/flow/cucumber/integration/ast/scans/AstRemoteRepoScanSteps.java:189]: The message should be changed from "scan ID" to "description".

File: src/test/java/com/checkmarx/flow/cucumber/integration/ast/scans/AstRemoteRepoScanSteps.java [hunk 5]
```diff
@@ -173,8 +178,30 @@ public void astScanWithInvalidAstMachine(String exceptionType, String errorSubst
...
+    public void startScanForInitiator(String initiatorList) {
+        List<VulnerabilityScanner> scanners = setScanInitiator(initiatorList);
+        startScan(scanners, BRANCH, PUBLIC_REPO);
+
+    }
+
+    private List<VulnerabilityScanner> setScanInitiator(String initiatorList) {
         String[] intiators ;
         List<VulnerabilityScanner> scanners = new LinkedList<>();
         if(!initiatorList.contains(SEPARATOR)){
```
Comment (alex-ko-dev) [inline src/test/java/com/checkmarx/flow/cucumber/integration/ast/scans/AstRemoteRepoScanSteps.java:207]: The whole statement can be replaced with initiatorList.split(SEPARATOR)

File: src/test/java/com/checkmarx/flow/cucumber/integration/azure/publishing/github2ado/Github2AdoSteps.java [hunk 6]
```diff
@@ -209,7 +210,14 @@ public void validateIssues(String project, String namespace){
             fail(e.getMessage());
         }
     }
-    
+
+    @And("description field is populated")
+    public void validateDescription() throws IOException {
+        azureDevopsClient.getIssues().forEach(issue -> {
```
Comment (alex-ko-dev) [inline src/test/java/com/checkmarx/flow/cucumber/integration/azure/publishing/github2ado/Github2AdoSteps.java:216]: Looks like an assertion is missing here.

File: src/test/resources/cucumber/features/integrationTests/ast/astScanProcessing.feature [hunk 7]
```diff
@@ -15,6 +15,19 @@ Feature: Cx-Flow AST Integration permutation tests
...
 
+  @ASTRemoteRepoScan @AdditionalFields
+  Scenario Outline: validate AST additional fields
+    Given enabled vulnerability scanners are "<scanners>"
+    Then scan results contain populated results for all scanners
+    And each finding will contain AST populated description field 
+    And finding with the same queryId will have the same description and there will be a unique finding description for each queryId  
+
+    Examples:
+    # Cannot rely on an exact number of findings, because it may change after backend version updates.
```
Comment (alex-ko-dev) [inline src/test/resources/cucumber/features/integrationTests/ast/astScanProcessing.feature:26]: This comment is not needed.

File: src/test/resources/cucumber/features/integrationTests/ast/astScanProcessing.feature [hunk 8]
```diff
@@ -15,6 +15,19 @@ Feature: Cx-Flow AST Integration permutation tests
...
+      | scanners | branch |
```
Comment (alex-ko-dev) [inline src/test/resources/cucumber/features/integrationTests/ast/astScanProcessing.feature:27]: The 'branch' column is not used.
The 'AST' value is only used once and can be put inline.

File: src/test/java/com/checkmarx/flow/cucumber/common/utils/AstUtils.java (hunk 2 above)
Comment (olgakil) [reply src/test/java/com/checkmarx/flow/cucumber/common/utils/AstUtils.java:18]: Hate java8

**Review by alex-ko-dev — APPROVED** [review]

**Pull request merged**